python/ray/dag/compiled_dag_node.py

```diff
@@ -1253,32 +1199,56 @@ def _get_or_compile(
-                # multiple readers on the same actor if the readers are on the same
-                # actor using IntraProcessChannel, which supports multiple readers
-                # on the same actor. We need to remove reader_handles_set in the future
-                # when we support multiple readers for both shared memory channel
-                # and IntraProcessChannel.
-                reader_handles_set = set()
+                # A dictionary that maps an InputNode or InputAttributeNode to its
+                # readers and the node on which the reader is running. Use `set` to
+                # deduplicate readers on the same actor because with CachedChannel
+                # each actor will only read from the shared memory once.
+                input_dag_node_to_reader_and_node_set: Dict[
+                    Union[InputNode, InputAttributeNode],
+                    Set[Tuple["ray.actor.ActorHandle", str]],
+                ] = defaultdict(set)
+
                 for idx in task.downstream_task_idxs:
                     reader_task = self.idx_to_task[idx]
                     assert isinstance(reader_task.dag_node, ClassMethodNode)
                     reader_handle = reader_task.dag_node._get_actor_handle()
-                    if reader_handle not in reader_handles_set:
-                        reader_and_node_list.append(
-                            (reader_handle, self._get_node_id(reader_handle))
-                        )
-                    reader_handles_set.add(reader_handle)
-                task.output_channels = [
-                    do_allocate_channel(
+                    reader_node_id = self._get_node_id(reader_handle)
+                    for arg in reader_task.args:
+                        if isinstance(arg, InputAttributeNode) or isinstance(
+                            arg, InputNode
+                        ):
+                            input_dag_node_to_reader_and_node_set[arg].add(
+                                (reader_handle, reader_node_id)
+                            )
+
+                # A single channel is responsible for sending the same data to
+                # corresponding consumers. Therefore, we create a channel for
+                # each InputAttributeNode, or a single channel for the entire
+                # input data if there are no InputAttributeNodes.
+                task.output_channels = []
+                for input_dag_node in input_dag_node_to_reader_and_node_set:
+                    reader_and_node_list = list(
+                        input_dag_node_to_reader_and_node_set[input_dag_node]
+                    )
+                    output_channel = do_allocate_channel(
                         self,
                         reader_and_node_list,
                         typ=type_hint,
                     )
-                ]
+                    task.output_channels.append(output_channel)
+                    task.output_idxs.append(
```

Comment:
not technically related to this PR, but why do we have separate list for output_idxs and output_channels? maybe we should combine it to a single dict

Reply:
I don't know, but I think using a list of tuples or a dictionary is better. Maybe we can revisit this when we have a more granular execution schedule generation mechanism. To generate a granular execution schedule, we should be able to control the order of `output_channels` in a writer.